### PR TITLE
[Map] add polyline support

### PR DESCRIPTION
**Pull request by sblondeau**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes 
| Issues        | Fix #... 
| License       | MIT

Add polyline support to Map for Leaflet and GoogleMap (useful e.g. for itinerary drawing)

```php
public function index(): Response
    {
        $map = (new Map('default'))
            ->center(new Point(45.7534031, 4.8295061))
            ->zoom(6)

            ->addPolyline(
                new Polyline(
                    title: 'my title',
                    points: [
                        new Point(48.8566, 2.3522),
                        new Point(45.7640, 4.8357),
                        new Point(43.2965, 5.3698),

                    ]
                )
            )
      ;          
        return $this->render('hom/index.html.twig', [
            'map' => $map,
        ]);;
    }
```

image

### CODE REVIEW

**Comment by WebMamba** [comment]
Hey, thanks a lot for this PR @sblondeau! Could you please add more info about this new feature to your PR description? Some screenshots, details and what you trying to achieve! Thannnnnks 😁

File: src/Map/doc/index.rst [hunk 1]
```diff
@@ -130,6 +130,22 @@ You can add markers to a map using the ``addMarker()`` method::
...
+        points: [
+            new Point(48.8566, 2.3522),
+            new Point(45.7640, 4.8357),
+            new Point(43.2965, 5.3698),
+            new Point(44.8378, -0.5792),
+        ],
+        infoWindow: new InfoWindow(
+            content: 'A line passing through Paris, Lyon, Marseille, Bordeaux',
+        ),
+    ));
```
Comment (Kocal) [inline src/Map/doc/index.rst]: Can you also move this section under `Add Polygons` please? Thanks :) 
```suggestion
You can add Polylines, which represents a path made by a series of ``Point`` instances::

    $myMap->addPolyline(new Polyline(
        points: [
            new Point(48.8566, 2.3522),
            new Point(45.7640, 4.8357),
            new Point(43.2965, 5.3698),
            new Point(44.8378, -0.5792),
        ],
        infoWindow: new InfoWindow(
            content: 'A line passing through Paris, Lyon, Marseille, Bordeaux',
        ),
    ));
```

File: src/Map/src/Bridge/Google/assets/src/map_controller.ts [hunk 2]
```diff
@@ -194,6 +219,23 @@ export default class extends AbstractMapController<
...
+                infoWindow.open({ map: this.map, anchor: element });
+            }
+        } else if (element instanceof google.maps.Polyline) {
+            element.addListener('click', (event: any) => {
+                if (definition.autoClose) {
+                    this.closeInfoWindowsExcept(infoWindow);
+                }
+                infoWindow.setPosition(event.latLng);
+                infoWindow.open(this.map);
+            });
```
Comment (Kocal) [inline src/Map/src/Bridge/Google/assets/src/map_controller.ts]: It looks like this logic is the same for Polygon, maybe we can merge the two blocks together?

File: src/Map/src/Twig/MapRuntime.php [hunk 3]
```diff
@@ -70,7 +76,7 @@ public function renderMap(
 
     public function render(array $args = []): string
     {
-        $map = array_intersect_key($args, ['map' => 0, 'markers' => 0, 'polygons' => 0, 'center' => 1, 'zoom' => 2]);
+        $map = array_intersect_key($args, ['map' => 0, 'markers' => 0, 'polygons' => 0, 'polylines' => 0, 'center' => 1, 'zoom' => 2]);
```
Comment (Kocal) [inline src/Map/src/Twig/MapRuntime.php]: I think there is an issue here, I don't fully remember why it has been done like this, but values should be sequential `0, 1, 2, 3, 4, ...`

**Review by Kocal — CHANGES_REQUESTED** [review]
Thanks for the feature, I didn't even know that "polyline" existed :)

I've added some comments that must be resolved if we want to merge your work.

Also, a lot of CI checks are failing, can you take a look please? Thanks!

File: src/Map/CHANGELOG.md [hunk 4]
```diff
@@ -10,6 +10,7 @@
 -   The importmap entry `@symfony/ux-map/abstract-map-controller` can be removed
     from your importmap, it is no longer needed. 
 -   Add `Polygon` support
+-   Add `Polyline` support
```
Comment (smnandre) [inline src/Map/CHANGELOG.md]: You can create a new section on top ? Next release will be 2.22 :)

File: src/Map/doc/index.rst [hunk 5]
```diff
@@ -148,7 +164,6 @@ You can also add Polygons, which represents an area enclosed by a series of ``Po
     ));
 
 Render a map
-------------
```
Comment (smnandre) [inline src/Map/doc/index.rst:157]: Why ? :)

File: src/Map/src/Bridge/Google/tests/GoogleRendererTest.php [hunk 6]
```diff
@@ -29,34 +29,34 @@ public function provideTestRenderMap(): iterable
...
+            'expected_render' => '<div data-controller="symfony--ux-google-map--map" data-symfony--ux-google-map--map-provider-options-value="{&quot;apiKey&quot;:&quot;api_key&quot;}" data-symfony--ux-google-map--map-view-value="{&quot;center&quot;:{&quot;lat&quot;:48.8566,&quot;lng&quot;:2.3522},&quot;zoom&quot;:12,&quot;fitBoundsToMarkers&quot;:false,&quot;options&quot;:{&quot;mapId&quot;:null,&quot;gestureHandling&quot;:&quot;auto&quot;,&quot;backgroundColor&quot;:null,&quot;disableDoubleClickZoom&quot;:false,&quot;zoomControlOptions&quot;:{&quot;position&quot;:22},&quot;mapTypeControlOptions&quot;:{&quot;mapTypeIds&quot;:[],&quot;position&quot;:14,&quot;style&quot;:0},&quot;streetViewControlOptions&quot;:{&quot;position&quot;:22},&quot;fullscreenControlOptions&quot;:{&quot;position&quot;:20}},&quot;markers&quot;:[],&quot;polygons&quot;:[];polylines&quot;:[]}"></div>',
```
Comment (smnandre) [inline src/Map/src/Bridge/Google/tests/GoogleRendererTest.php]: Could we not send empty values ? i fear this will quickly become a long list 

cc @Kocal

File: src/Map/src/Polyline.php [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+ *
+ * @author [Pierre Svgnt]
+ */
+final readonly class Polyline
+{
+    /**
+     * @param array<string, mixed> $extra Extra data, can be used by the developer to store additional information and use them later JavaScript side
+     */
+    public function __construct(
+        private array $points,
```
Comment (smnandre) [inline src/Map/src/Polyline.php:27]: add docblock to type the array

File: src/Map/src/Polyline.php [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+ */
+
+namespace Symfony\UX\Map;
+
+use Symfony\UX\Map\Exception\InvalidArgumentException;
+
+/**
+ * Represents a polyline on a map.
+ *
+ * @author [Pierre Svgnt]
```
Comment (smnandre) [inline src/Map/src/Polyline.php]: Are you sure ? :)

File: src/Map/src/Polyline.php [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+        private ?string $title = null,
+        private ?InfoWindow $infoWindow = null,
+        private array $extra = [],
+    ) {
+    }
+
+    /**
+     * Convert the polyline to an array representation.
+     */
+    public function toArray(): array
```
Comment (smnandre) [inline src/Map/src/Polyline.php]: Let's add a docblock here too (see Marker.php for example)

File: src/Map/src/Twig/MapRuntime.php (hunk 3 above)
Comment (smnandre) [reply src/Map/src/Twig/MapRuntime.php]: No need here, as it's then using named arguments with the spread.
 
 But it's a good idea at least to prevent future errors if we change to some array_reverse :)

File: src/Map/src/Twig/UXMapComponent.php [hunk 10]
```diff
@@ -35,4 +36,8 @@ final class UXMapComponent
      * @var Polygon[]
      */
     public array $polygons;
+    /**
+     * @var Polyline[]
```
Comment (smnandre) [inline src/Map/src/Twig/UXMapComponent.php:41]:
```suggestion

    /**
     * @var Polyline[]
```

File: src/Map/src/Bridge/Google/tests/GoogleRendererTest.php (hunk 6 above)
Comment (Kocal) [reply src/Map/src/Bridge/Google/tests/GoogleRendererTest.php]: It's a topic I have in mind for a moment now, let's do that later

Comment (smnandre) [reply src/Map/src/Bridge/Google/tests/GoogleRendererTest.php]: Of course :)

File: src/Map/src/Bridge/Google/assets/dist/map_controller.js [hunk 11]
```diff
@@ -1,69 +1,8 @@
-import { Controller } from '@hotwired/stimulus';
+import AbstractMapController from '@symfony/ux-map';
```
Comment (smnandre) [inline src/Map/src/Bridge/Google/assets/dist/map_controller.js]: What happened here ? 

ping @Kocal

File: src/Autocomplete/assets/dist/controller.js [hunk 12]
```diff
@@ -15,7 +15,7 @@ LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
 OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
 PERFORMANCE OF THIS SOFTWARE.
 ***************************************************************************** */
-/* global Reflect, Promise, SuppressedError, Symbol */
+/* global Reflect, Promise, SuppressedError, Symbol, Iterator */
```
Comment (smnandre) [inline src/Autocomplete/assets/dist/controller.js]: Let's revert this change (or see in another PR) to focus on one package :)

**Comment by sblondeau** [comment]
Hello, I tried a lot of things but some tests (js ans php) failed and I do not understand why :( 
If someone has some time to investigate... 
Thanks :-)

**Comment by Kocal** [comment]
Hi @sblondeau, where are stuck?

Please ensure to rebase your branch from the latest `2.x` quite often, but also try to run some yarn scripts to easy your development:
```shell
yarn lint
yarn format
yarn workspaces foreach -Rp --from '{src/Map/assets,src/Map/src/Bridge/Google/assets,src/Map/src/Bridge/Leaflet/assets}' run build
yarn workspaces foreach -Rp --from '{src/Map/assets,src/Map/src/Bridge/Google/assets,src/Map/src/Bridge/Leaflet/assets}' run test
```

Also, please start to squash your commits, we don't want to see commits [`1517fe1` (#2340)](https://github.com/symfony/ux/pull/2340/commits/1517fe1c303cabfe913236b8fe470a37062b0223) like this because it will _spam_ the history for nothing (meaning, harder to blame if necessary).

Thanks!

**Comment by sblondeau** [comment]
@Kocal thanks for your help with the yarn command (it is not easy to guess them ;-) )
My code works loclaly but I do not understand how to correctly launch tests. I have a failing test with GoogleMap. Sorry if my question sounds stupid but I do not understand how to correctly contribute with this part, should I work in ux map repo or in the bridge repo ? It is frustrating to be stuck so close to the end ;-) Thanks for your help

**Comment by Kocal** [comment]
> @Kocal thanks for your help with the yarn command (it is not easy to guess them ;-) )

Well, in fact
```
yarn workspaces foreach -Rp --from '{src/Map/assets,src/Map/src/Bridge/Google/assets,src/Map/src/Bridge/Leaflet/assets}' run build
``` 
is simply a shortcut to:
```
(cd src/Map/assets && yarn build)
(cd src/Map/src/Bridge/Google/assets && yarn build)
(cd src/Map/src/Bridge/Leaflet/assets && yarn build)
``` 
😛 


> My code works loclaly but I do not understand how to correctly launch tests. I have a failing test with GoogleMap. Sorry if my question sounds stupid but I do not understand how to correctly contribute with this part, should I work in ux map repo or in the bridge repo ?

Yes! You should work in the UX repository, `symfony/ux-*` are supposed to be read-only, but it looks like it is not the case. 

To run JS tests for Google, you can run `cd src/Map/src/Bridge/Google/assets && yarn test`.

Keep up the good work! But... I predict you will have a lot of conflicts when #2385 will be merged :P

File: src/Map/doc/index.rst [hunk 13]
```diff
@@ -41,6 +41,12 @@ Configuration is done in your ``config/packages/ux_map.yaml`` file:
             # without to manually configure it in each map instance (through "new GoogleOptions(mapId: 'your_map_id')").
             default_map_id: null
 
+        # Google Maps specific configuration
+        google_maps:
+            # Configure the default Map Id (https://developers.google.com/maps/documentation/get-map-id),
+            # without to manually configure it in each map instance (through "new GoogleOptions(mapId: 'your_map_id')").
+            default_map_id: null
+
 The ``UX_MAP_DSN`` environment variable configure which renderer to use.
```
Comment (smnandre) [inline src/Map/doc/index.rst]: ??

**Review by Kocal — CHANGES_REQUESTED** [review]
Please can you squash all of your commits and rebase on the latest 2.x?
Thanks

**Comment by Kocal** [comment]
With the merge of #2385 , there are now plenty of conflicts 😅 
image


I'll take it from here :)

**Comment by Kocal** [comment]
PR squashed and rebased, I've added compatibility with UX Map inside a LiveComponent (I also fixed some things that I've forgot in #2385 😅), and applied some review comments.

**Review by Kocal — APPROVED** [review]

**Review by smnandre — APPROVED** [review]
Well, @sblondeau, what a nice first contribution here! 

Thank you very much, and i cannot wait to see the next ones :)

**Comment by smnandre** [comment]
Thanks for your work on this new feature!
